File: leetcode/457-Circular_Array_Loop/main.py
class Solution:
    def circularArrayLoop(self, nums: List[int]) -> bool:
        for i in range(len(nums)):
            # True if positive else False
            pre_direction = nums[i] > 0
            slow_i = fast_i = i

            while True:
                # Move slow_i
                slow_i = self._next_idx(slow_i, nums[slow_i], len(nums))
                if self._is_not_cycle(pre_direction, nums[slow_i], len(nums)):
                    break
                # Move fast_i
                fast_i = self._next_idx(fast_i, nums[fast_i], len(nums))
                if self._is_not_cycle(pre_direction, nums[fast_i], len(nums)):
                    break
                # Move fast_i again
                fast_i = self._next_idx(fast_i, nums[fast_i], len(nums))
                if self._is_not_cycle(pre_direction, nums[fast_i], len(nums)):
                    break
                # The direction has changed
                if slow_i == fast_i:
                    return True

        return False

    def _next_idx(self, index: int, value: int, size: int) -> int:
        result = (index + value) % size
        if result < 0:
            result = result + size
        return result

    def _is_not_cycle(
        self, pre_direction: bool, curr_value: int, size_of_nums: int
    ) -> bool:
        # True if positive else False
        curr_direction = curr_value > 0
        if pre_direction != curr_direction or curr_value % size_of_nums == 0:
            return True
        return False

File: leetcode/457-Circular_Array_Loop/main.go
package circulararrayloop

func circularArrayLoop(nums []int) bool {
	size := len(nums)
	// for idx, val := range nums {
	for i := 0; i < len(nums); i++ {
		slow := i
		fast := i

		direction := nums[i] >= 0

		for {
			slow = moveTo(nums[slow], slow, size)
			if isNotCycle(nums, direction, slow) {
				break
			}

			fast = moveTo(nums[fast], fast, size)
			if isNotCycle(nums, direction, fast) {
				break
			}
			fast = moveTo(nums[fast], fast, size)
			if isNotCycle(nums, direction, fast) {
				break
			}

			if slow == fast {
				return true
			}
		}
	}
	return false
}

// val is the value of nums[idx]
// idx is the index of nums
// size is the length of nums
func moveTo(val, idx, size int) int {
	result := (val + idx) % size
	if result < 0 {
		result += size
	}
	return result
}

func isNotCycle(nums []int, prevDirection bool, idx int) bool {
	currentDirection := nums[idx] >= 0
	if (currentDirection != prevDirection) || (nums[idx]%len(nums) == 0) {
		return true
	}
	return false
}
